### PR TITLE
Canonicalize order of If and Loop outputs

### PR DESCRIPTION
**Pull request by eellison**
Canonicalize the ordering of outputs of if and loop nodes based on their first usage. Previously we were able to canonicalize output order by sorting on variable name, but this breaks down with outputs added in an early return pass.

### CODE REVIEW

**Comment by eellison** [comment]
We might consider only running this before exporting, so as to not add the compilation overhead when it's not needed

File: torch/csrc/jit/passes/canonicalize.cpp [hunk 1]
```diff
@@ -47,5 +48,220 @@ std::shared_ptr<Graph> Canonicalize(
...
+void swapIfBlockOutputs(Block* b, const std::vector<size_t>& new_indices) {
+  for (size_t index : new_indices) {
+    b->registerOutput(b->outputs().at(index));
+  }
+  for (size_t i = 0; i < new_indices.size(); ++i) {
+    b->eraseOutput(0);
+  }
+}
+
+void swapLoopBlockOutputs(Node* n, const std::vector<size_t>& new_indices) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/canonicalize.cpp]: This code is hard to follow because it is combining a couple things into one:

1. Permuting inputs or outputs to a function.
2. Keeping track of the extra non-permuted inputs

Instead can we:

1. add permuteInputs/permuteOutputs directly to ir.h (If you look at what removeInput does, it is not actually that hard, it just requires adjusting the Use objects offsets).  This avoids having all the logic to copy meta-data because we are just swapping the nodes
2. Add permuteLoopCarried and permuteOutputs to LoopView, IfView, because they are the only classes that should know about carried_inputs_offsets, etc.
3. Redo this code to just call the  view versions.

**Review by zdevito — COMMENTED** [review]
General structure looks fine. A few comments about correctness and readability.

File: torch/csrc/jit/passes/canonicalize.cpp [hunk 2]
```diff
@@ -47,5 +48,220 @@ std::shared_ptr<Graph> Canonicalize(
...
+}
+
+std::vector<size_t> sort_indexes(at::ArrayRef<Value*> values) {
+  // initialize original index locations
+  std::vector<size_t> idx(values.size());
+  std::iota(idx.begin(), idx.end(), 0);
+
+  // Sort values based on canonical ordering of their first usage
+  std::sort(idx.begin(), idx.end(), [&values](size_t i1, size_t i2) {
+    if (values[i1]->uses().size() == 0 && values[i2]->uses().size() == 0) {
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/canonicalize.cpp]: See added comment, should be fine

File: torch/csrc/jit/passes/canonicalize.cpp [hunk 3]
```diff
@@ -47,5 +48,220 @@ std::shared_ptr<Graph> Canonicalize(
...
+      return i1 < i2;
+    }
+    if (values[i1]->uses().size() == 0) {
+      return false;
+    } else if (values[i2]->uses().size() == 0) {
+      return true;
+    }
+
+    auto fst_v1 = values[i1]->uses()[0];
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/canonicalize.cpp]: i think this is fine to use ? i think it's stable

**Comment by eellison** [comment]
@zdevito take another look?

File: torch/csrc/jit/passes/canonicalize.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/passes/canonicalize.cpp]: But it isn't the first use. It might take carefully constructing a test case to make this so, but there is nothing keeping them in order. That is only one step away from causing this pass to break under extremely hard-to-understand circumstances.

File: torch/csrc/jit/ir.cpp [hunk 4]
```diff
@@ -1138,6 +1138,31 @@ void Node::removeAllInputs() {
   inputs_.clear();
 }
 
+void Node::permuteInputs(const std::vector<size_t>& new_order) {
+  schema_ = nullptr;
+  AT_ASSERT(new_order.size() == inputs_.size());
+  std::vector<Value*> new_inputs;
+  new_inputs.reserve(new_order.size());
+  for (size_t i = 0; i < new_order.size(); ++i) {
+    new_inputs.push_back(inputs_[new_order[i]]);
```
Comment (zdevito) [inline torch/csrc/jit/ir.cpp]: Please use `.at` on all vector lookups into new_order, otherwise this will segfault when bad indices.

File: torch/csrc/jit/ir.cpp [hunk 5]
```diff
@@ -1138,6 +1138,31 @@ void Node::removeAllInputs() {
...
+    auto it = findUseForInput(new_order[i]);
+    it->offset = i;
+  }
+  inputs_ = std::move(new_inputs);
```
Comment (zdevito) [inline torch/csrc/jit/ir.cpp]: In both loops, can you assert that a value is not being used twice. This can be done by blanking out the Value* in the original list, and checking that the value is not-null on lookup.

File: torch/csrc/jit/ir.h [hunk 6]
```diff
@@ -554,6 +554,13 @@ struct TORCH_API Node {
   // Result: %3 = f()
   void removeAllInputs();
 
+  // Rearrange the ordering of inputs or outputs of a node
```
Comment (zdevito) [inline torch/csrc/jit/ir.h]: comment: each index must appear exactly once.

File: torch/csrc/jit/ir_views.h [hunk 7]
```diff
@@ -80,9 +86,33 @@ struct LoopView {
     return node_;
   }
 
+  // adjust index_ordering by adding indices 0 - thorugh adjust, and
+  // incrementing all existing inputs by adjust
+  std::vector<size_t> adjustIndices(
```
Comment (zdevito) [inline torch/csrc/jit/ir_views.h]: nit: this is not a method of LoopView because it doesn't use any members of loop view.

File: torch/csrc/jit/passes/canonicalize.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/passes/canonicalize.cpp]: Pretty sure this is a real bug. It hasn't been hit yet because nothing has transformed the graph enough to make the Uses get out of sync with the  order in the graph, but it is a bug waiting to happen.

**Review by zdevito — APPROVED** [review]
Ready to go once uses[0] is fixed. Other issues are easy to fix.

File: torch/csrc/jit/ir.cpp (hunk 5 above)
Comment (eellison) [reply torch/csrc/jit/ir.cpp]: I can check that an index is not being used twice, but since this is being used on node inputs a value can appear multiple times
